src/Resolver/ResourceToTeaserResolver.php

```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Atoolo\GraphQL\Search\Resolver;
+
+use Atoolo\GraphQL\Search\Types\ArticleTeaser;
+use Atoolo\GraphQL\Search\Types\MediaTeaser;
+use Atoolo\GraphQL\Search\Types\Teaser;
+use Atoolo\Resource\Resource;
+use Overblog\GraphQLBundle\Definition\ArgumentInterface;
+
+class ResourceToTeaserResolver
+{
+    /**
+     * @param iterable<TeaserResolver> $resolverList
+     */
+    public function __construct(
+        private readonly iterable $resolverList
+    ) {
+    }
+
+    public function resolve(Resource $resource): Teaser
+    {
+        foreach ($this->resolverList as $resolver) {
+            if (!$resolver->accept($resource)) {
+                continue;
+            }
+            return $resolver->resolve($resource);
+        }
+
+        throw new \InvalidArgumentException(
+            'Unresolvable resource ' . $resource->getLocation()
+        );
+    }
```

Comment:
So I've thought about this a bit.  
I dislike the separation between `accept` and `resolve`. How about we combine them?
```php
        foreach ($this->factories as $factory)  {
            $result = $factory->tryCreate($resource);
            if ($result instanceof NotResponsible) {
                continue;
            }
            if ($result instanceof Erroneous) {
                throw $result->getError();
            }
            // last check is only for the sake of demonstration 
            if ($result instanceof Successfull) {
                return $result->getTeaser();
            }
        }
```
We'd have a `Resolver/TeaserFactory/Result/` namespace containing these three classes and the `TeaserFactory` interface would consist of this method:
```php
    public function tryCreate(
        Resource $resource
    ): Successfull|NotResponsible|Erroneous;
```

I don't know about iterating all and having priorities defined in the di config. from experience this usually ends up with the default (`ArticleTeaserResolver` in this case) having a lower priority and all others beein the same. And since we stop after the first success it is the developers responsibility to define all `TeaserResolver`s in such a way, that no more than one (appart from the default) can `accept` any `Resource`.  
So far I have not come to a (practical) solution. For now this will (have to) work.

Reply:
I think that's a bit overengineered. Wouldn't you do it like this in PHP?

```php
        foreach ($this->factories as $factory)  {
            $result = $factory->create($resource);
            if ($result === false) {
                continue;
            }
            return $result->getTeaser();
        }
        throw new \InvalidArgumentException(
            'Unresolvable resource ' . $resource->getLocation()
        );
```

```php
    public function create(
        Resource $resource
    ): false|Teaser;
```

Reply:
Yea, on second thought that's probably true...

What about we do it similar to what's planed for the `HierarchyResolver` via `index_by`?  
We would register all `TeaserFactory`s (if renamed to this, currently called `TeaserResolver`) via di where the `index` is the `objectType` we want it to be invoked for + a "Default" or "Fallback" Factory to handle all other cases.  

This would solve multiple issues I had initially:
- `TeaserFactory`s do not have to decide weather or not they can use a given `Resource`
- they have one job and one job only: the creation of a `Teaser` from a `Resource`
- No iterating with arbitrary priorities is required. So we don't have to be warry of `Factory`s beeing able to `accept` `Resources` others might `accept` too.
- it's more straight forward to replace pre-configured `Factory`s in the di config

Reply:
Yes, I think that's a good idea. I've refactored it.

Reply:
looks great, nice work!